File: TaskScheduler.java
/*
* Did this code successfully run on Leetcode : YES
* 
* Any problem you faced while coding this : NO
* 
* Time Complexity: O(n)
* 
* Space Complexity: O(1)
* 
*/

import java.util.HashMap;

public class TaskScheduler {
    public int leastInterval(char[] tasks, int n) {
        HashMap<Character, Integer> hmap = new HashMap<>();

        int maxFreq = 0;

        for (char task : tasks) {
            hmap.put(task, hmap.getOrDefault(task, 0) + 1);

            maxFreq = Math.max(maxFreq, hmap.get(task));
        }

        int sameFreqElements = 0;

        for (char key : hmap.keySet()) {
            if (hmap.get(key) == maxFreq) {
                sameFreqElements++;
            }
        }

        int partitions = maxFreq - 1;

        int availableSlots = partitions * (n - (sameFreqElements - 1));

        int pendingTasks = tasks.length - maxFreq * sameFreqElements;

        int idleSlots = Math.max(0, availableSlots - pendingTasks);

        return tasks.length + idleSlots;
    }
}

File: task-scheduler.py
#Time Complexity : O(N)
#Space Complexity : O(1) : constrained by finite number of alphabets


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        
        task_freq = {}
        for task in tasks:
            if task in task_freq:
                task_freq[task] += 1
            else:
                task_freq[task] = 1

        max_freq = max(task_freq.values())
        tasks_with_max_freq = sum(freq == max_freq for freq in task_freq.values())

        partitions = max_freq - 1
        available_slots = partitions * (n - (tasks_with_max_freq - 1))
        pending_tasks = len(tasks) - (max_freq * tasks_with_max_freq)
        idle_slots = max(0, available_slots - pending_tasks)

        return len(tasks) + idle_slots
